velox/functions/sparksql/DecimalArithmetic.cpp

```diff
@@ -14,379 +14,71 @@
 
 namespace facebook::velox::functions::sparksql {
 namespace {
 
-std::string getResultScale(std::string precision, std::string scale) {
-  return fmt::format(
-      "({}) <= 38 ? ({}) : max(({}) - ({}) + 38, min(({}), 6))",
-      precision,
-      scale,
-      scale,
-      precision,
-      scale);
-}
-
-// Returns the whole and fraction parts of a decimal value.
-template <typename T>
-inline std::pair<T, T> getWholeAndFraction(T value, uint8_t scale) {
-  const auto scaleFactor = velox::DecimalUtil::kPowersOfTen[scale];
-  const T whole = value / scaleFactor;
-  return {whole, value - whole * scaleFactor};
-}
-
-// Increases the scale of input value by 'delta'. Returns the input value if
-// delta is not positive.
-inline int128_t increaseScale(int128_t in, int16_t delta) {
-  // No need to consider overflow as 'delta == higher scale - input scale', so
-  // the scaled value will not exceed the maximum of long decimal.
-  return delta <= 0 ? in : in * velox::DecimalUtil::kPowersOfTen[delta];
-}
-
-// Scales up the whole part to result scale, and combine it with fraction part
-// to produce a full result for decimal add. Checks whether the result
-// overflows.
-template <typename T>
-inline T
-decimalAddResult(T whole, T fraction, uint8_t resultScale, bool& overflow) {
-  T scaledWhole = DecimalUtil::multiply<T>(
-      whole, velox::DecimalUtil::kPowersOfTen[resultScale], overflow);
-  if (FOLLY_UNLIKELY(overflow)) {
-    return 0;
-  }
-  const auto result = scaledWhole + fraction;
-  if constexpr (std::is_same_v<T, int64_t>) {
-    overflow = (result > velox::DecimalUtil::kShortDecimalMax) ||
-        (result < velox::DecimalUtil::kShortDecimalMin);
-  } else {
-    overflow = (result > velox::DecimalUtil::kLongDecimalMax) ||
-        (result < velox::DecimalUtil::kLongDecimalMin);
-  }
-  return result;
-}
+struct BinaryPrecisionScales {
+  uint8_t aPrecision;
+  uint8_t aScale;
+  uint8_t bPrecision;
+  uint8_t bScale;
+};
 
-// Reduces the scale of input value by 'delta'. Returns the input value if delta
-// is not positive.
-template <typename T>
-inline static T reduceScale(T in, int32_t delta) {
-  if (delta <= 0) {
-    return in;
-  }
-  T result;
-  bool overflow;
-  const auto scaleFactor = velox::DecimalUtil::kPowersOfTen[delta];
-  if constexpr (std::is_same_v<T, int64_t>) {
-    VELOX_DCHECK_LE(
-        scaleFactor,
-        std::numeric_limits<int64_t>::max(),
-        "Scale factor should not exceed the maximum of int64_t.");
-  }
-  DecimalUtil::divideWithRoundUp<T, T, T>(
-      result, in, T(scaleFactor), 0, overflow);
-  VELOX_DCHECK(!overflow);
+// Extracts the precisions and scales for two decimal types.
+BinaryPrecisionScales getDecimalPrecisionScales(
+    const TypePtr& aType,
+    const TypePtr& bType) {
+  auto [aPrecision, aScale] = getDecimalPrecisionScale(*aType);
+  auto [bPrecision, bScale] = getDecimalPrecisionScale(*bType);
+  BinaryPrecisionScales result{
+      static_cast<uint8_t>(aPrecision),
```

Comment:
> static_cast<uint8_t>

Why do we use static_cast? Is it needed? Should we change the signature of getDecimalPrecisionScale instead?

Reply:
Spark supports negative scale, so I assume we don't change the function return type, now it is int.

Reply:
And some computation needs the precision and scale to be int.

Reply:
> Spark supports negative scale, so I assume we don't change the function return type, now it is int.

I didn't know that. Do we have tests for negative scale? How this this cast work in that case? CC: @rui-mo

Reply:
We don't test for negative scale, now we fallback all the negative scale for decimal functions in gluten.

Reply:
In that case, assuming Velox doesn't support negative precision / scale, can we change getDecimalPrecisionScale signature or remove these static casts?